pkg/operator/controller/gatewayclass/subscription.go

```diff
@@ -0,0 +1,120 @@
+func (r *reconciler) ensureServiceMeshOperatorSubscription(ctx context.Context) (bool, *operatorsv1alpha1.Subscription, error) {
+	name := types.NamespacedName{
+		// TODO Use named consts for the namespace and name.
+		Namespace: "openshift-operators",
+		Name:      "servicemeshoperator",
+	}
+	have, current, err := r.currentSubscription(ctx, name)
+	if err != nil {
+		return false, nil, err
+	}
+
+	desired, err := desiredSubscription(name)
+	if err != nil {
+		return have, current, err
+	}
+
+	switch {
+	case !have:
+		if err := r.createSubscription(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentSubscription(ctx, name)
+	case have:
+		if updated, err := r.updateSubscription(ctx, current, desired); err != nil {
+			return have, current, err
+		} else if updated {
+			return r.currentSubscription(ctx, name)
+		}
+	}
+	return true, current, nil
+}
+
+// desiredSubscription returns the desired subscription.
+func desiredSubscription(name types.NamespacedName) (*operatorsv1alpha1.Subscription, error) {
+	subscription := operatorsv1alpha1.Subscription{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: name.Namespace,
+			Name:      name.Name,
+		},
+		Spec: &operatorsv1alpha1.SubscriptionSpec{
+			// TODO: Use named consts if available.
+			Channel:                "stable",
+			InstallPlanApproval:    operatorsv1alpha1.ApprovalAutomatic,
+			Package:                "servicemeshoperator",
+			CatalogSource:          "redhat-operators",
+			CatalogSourceNamespace: "openshift-marketplace",
+		},
+	}
+	return &subscription, nil
+}
+
+// currentSubscription returns the current subscription.
+func (r *reconciler) currentSubscription(ctx context.Context, name types.NamespacedName) (bool, *operatorsv1alpha1.Subscription, error) {
+	var subscription operatorsv1alpha1.Subscription
+	if err := r.client.Get(ctx, name, &subscription); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, fmt.Errorf("failed to get subscription %s: %w", name, err)
+	}
+	return true, &subscription, nil
+}
+
+// createSubscription creates a subscription.
+func (r *reconciler) createSubscription(ctx context.Context, subscription *operatorsv1alpha1.Subscription) error {
+	if err := r.client.Create(ctx, subscription); err != nil {
+		return fmt.Errorf("failed to create subscription %s/%s: %w", subscription.Namespace, subscription.Name, err)
+	}
+	log.Info("created subscription", "namespace", subscription.Namespace, "name", subscription.Name)
+	return nil
+}
+
+// updateSubscription updates a subscription.
+func (r *reconciler) updateSubscription(ctx context.Context, current, desired *operatorsv1alpha1.Subscription) (bool, error) {
+	changed, updated := subscriptionChanged(current, desired)
+	if !changed {
+		return false, nil
+	}
+
+	// Diff before updating because the client may mutate the object.
+	diff := cmp.Diff(current, updated, cmpopts.EquateEmpty())
```

Comment:
It doesn't look like like anything in this `Subscription` would ever change even if the version changed.  I wonder if there needs to be another field set in `desiredSubscription`, e.g. `StartingCSV` ?